node/node.go

```diff
@@ -1775,6 +1784,64 @@ func (app *App) setupDBs(ctx context.Context, lg log.Log) error {
 	return nil
 }
 
+// MigrateLocalDB migrates the old node_state.sql to the new local.sql
+//
+// This function is idempotent and can be called multiple times without side effects.
+// It will only migrate the old db to the new db if the old db exists and the new db does not.
+//
+// TODO(mafa): this can be removed in the future when we are sure that all nodes have migrated to the new db.
+func (app *App) MigrateLocalDB(lg *zap.Logger, dbPath string, clients []localsql.PoetClient) error {
+	oldDBFile := filepath.Join(dbPath, oldLocalDbFile)
+	dbFile := filepath.Join(dbPath, localDbFile)
+	_, err := os.Stat(oldDBFile)
+	switch {
+	case errors.Is(err, fs.ErrNotExist):
+		return nil // no old db to migrate
+	case err != nil:
+		return fmt.Errorf("stat %s: %w", oldDBFile, err)
+	}
+
+	_, err = os.Stat(dbFile)
+	switch {
+	case errors.Is(err, fs.ErrNotExist):
+		// no new db, migrate old to new
+	case err == nil:
+		// both exist, error
+		return fmt.Errorf("%w: both %s and %s exist", fs.ErrExist, oldDBFile, dbFile)
+	case err != nil:
+		return fmt.Errorf("stat %s: %w", dbFile, err)
+	}
+
+	lg.Info("migrating local DB",
+		zap.String("old db", oldDBFile),
+		zap.String("new db", dbFile),
+	)
+	migrations, err := sql.LocalMigrations()
+	if err != nil {
+		return fmt.Errorf("load local migrations: %w", err)
+	}
+	oldDB, err := localsql.Open("file:"+oldDBFile,
+		sql.WithLogger(lg),
+		sql.WithMigrations(migrations),
+		sql.WithMigration(localsql.New0001Migration(app.Config.SMESHING.Opts.DataDir)),
+		sql.WithMigration(localsql.New0002Migration(app.Config.SMESHING.Opts.DataDir)),
+		sql.WithMigration(localsql.New0003Migration(app.Config.SMESHING.Opts.DataDir, clients)),
+		sql.WithConnections(app.Config.DatabaseConnections),
+	)
+	if err != nil {
+		return fmt.Errorf("open sqlite db %w", err)
+	}
+	defer oldDB.Close()
+
+	if _, err := oldDB.Exec(fmt.Sprintf("VACUUM INTO '%s'", dbFile), nil, nil); err != nil {
+		return fmt.Errorf("vacuum %s to %s: %w", oldDBFile, dbFile, err)
+	}
```

Comment:
Would it make sense to use https://pkg.go.dev/github.com/go-llsqlite/llsqlite#Conn.BackupToDB?

Reply:
Theoretically yes, but VACUUM does the same thing, with the added benefit of cleaning up / shrinking the DB in the process.